### PR TITLE
KTON Staking Part.1

### PR DESCRIPTION
**Pull request by AurevoirXavier**

### CODE REVIEW

File: pallet/staking/src/lib.rs [hunk 1]
```diff
@@ -1208,3 +1225,55 @@ where
...
+}
+impl KtonStakerNotification for () {}
+/// KTON staker contact notifier.
+pub struct KtonStakerNotifier<T>(PhantomData<T>);
+impl<T> KtonStakerNotification for KtonStakerNotifier<T>
+where
+	T: darwinia_message_transact::Config,
+{
+	fn notify() {
+		// if let Err(e) = darwinia_message_transact::Pallet::message_transact() {
```
Comment (AurevoirXavier) [inline pallet/staking/src/lib.rs:1275]: Please help construct the notify transact. @boundless-forest @hujw77

**Comment by AurevoirXavier** [comment]
Please help fix the precompile test. @boundless-forest

**Comment by boundless-forest** [comment]
> Please help fix the precompile test. @boundless-forest

I didn't have a deep understanding of the current KTON design, so just fix the precompile tests according to your code changes. It seems that only one interface has been updated.

File: pallet/staking/src/lib.rs [hunk 2]
```diff
@@ -351,6 +370,12 @@ pub mod pallet {
 	pub struct Pallet<T>(_);
 	#[pallet::hooks]
 	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_runtime_upgrade() -> Weight {
+			<MigrationStartBlock<T>>::put(<frame_system::Pallet<T>>::block_number());
+
+			T::DbWeight::get().reads_writes(0, 1)
```
[the remark this answers is not in the document]
Comment (AurevoirXavier) [reply pallet/staking/src/lib.rs:376]: block number is a whitelist storage.

File: pallet/staking/src/lib.rs [hunk 3]
```diff
@@ -451,16 +476,12 @@ pub mod pallet {
...
 						ring_amount,
 					)?;
 				}
 				if kton_amount != 0 {
-					Self::unstake_token::<KtonPool<T>>(
-						&mut l.staked_kton,
-						&mut l.unstaking_kton,
-						kton_amount,
-					)?;
+					Self::unstake_token::<KtonPool<T>>(&mut l.staked_kton, None, kton_amount)?;
```
Comment (AurevoirXavier) [inline pallet/staking/src/lib.rs:484]: transfer back

File: pallet/staking/src/lib.rs [hunk 4]
```diff
@@ -883,42 +903,43 @@ pub mod pallet {
...
 				.unwrap_or_default()
 		}
 
 		/// Distribute the session reward to staking pot and update the stakers' reward record.
-		pub fn distribute_session_reward(amount: Balance) -> Balance {
+		pub fn distribute_session_reward(amount: Balance) {
+			let (reward_to_v1, reward_to_v2) = {
+				let reward_to_ring = amount / 2;
+				let reward_to_kton = amount - reward_to_ring;
+				let reward_to_kton_v1 = T::MigrationCurve::get() * reward_to_kton;
```
Comment (AurevoirXavier) [inline pallet/staking/src/lib.rs:917]:
```
y = 100 - (100 / x) * t
```

File: pallet/staking/src/lib.rs [hunk 5]
```diff
@@ -883,42 +903,43 @@ pub mod pallet {
...
-				staker: staking_pot.clone(),
-				amount: actual_reward,
-			});
+			};
 
-			if unpaid != 0 {
-				Self::deposit_event(Event::Unpaid { staker: staking_pot, amount: unpaid });
-			}
+			reward(staking_pot, actual_reward_v1);
+			reward(T::KtonStakerAddress::get(), reward_to_v2);
```
Comment (AurevoirXavier) [inline pallet/staking/src/lib.rs:940]: distribution contract address

Comment (hujw77) [reply pallet/staking/src/lib.rs:941]: https://github.com/darwinia-network/kton-staker?tab=readme-ov-file#addresses
```
|  Contract            | Canonical Cross-chain Deployment Address   |
|----------------------|--------------------------------------------|
|  KTONStakingRewards  | 0x000000000419683a1a03AbC21FC9da25fd2B4dD7 |
|  RewardsDistribution | 0x000000000Ae5DB7BDAf8D071e680452e33d91Dd5 |
|  sc/ktstk            | 0x73632F6B7473746B000000000000000000000000 |
```

**Comment by hackfisher** [comment]
@AurevoirXavier Still some issues with CI.

File: precompile/staking/src/lib.rs [hunk 6]
```diff
@@ -97,11 +97,10 @@ where
 		Ok(true)
 	}
 
-	#[precompile::public("restake(uint256,uint256,uint8[])")]
+	#[precompile::public("restake(uint256,uint8[])")]
```
Comment (hackfisher) [inline precompile/staking/src/lib.rs:100]: @JayJay1024 Did we support restake on UI? Need to adapt if the answer is yes.

File: pallet/staking/src/lib.rs [hunk 7]
```diff
@@ -883,42 +909,47 @@ pub mod pallet {
...
-		pub fn distribute_session_reward(amount: Balance) -> Balance {
+		pub fn distribute_session_reward(amount: Balance) {
+			let (reward_to_v1, reward_to_v2) = {
+				let reward_to_ring = amount / 2;
+				let reward_to_kton = amount - reward_to_ring;
+				#[cfg(not(any(test, feature = "runtime-benchmarks")))]
+				let ratio = migration_curve_kton_reward(T::MigrationCurve::get());
+				#[cfg(any(test, feature = "runtime-benchmarks"))]
+				let ratio = Perquintill::one();
+				let reward_to_kton_v1 = ratio * reward_to_kton;
```
Comment (hackfisher) [inline pallet/staking/src/lib.rs:927]: I'm sorry, my math calculation was incorrect, here ratio should be x/100

File: pallet/staking/src/lib.rs [hunk 8]
```diff
@@ -1208,3 +1235,136 @@ where
...
+/// ```
+/// b"sc/ktstk"
+/// ```
+pub struct KtonStakerAddress;
+impl<T> Get<T> for KtonStakerAddress
+where
+	T: From<[u8; 20]>,
+{
+	fn get() -> T {
+		[115, 99, 47, 107, 116, 115, 116, 107, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0].into()
```
[the remark this answers is not in the document]
Comment (hujw77) [reply pallet/staking/src/lib.rs:1249]: It is right.
```
~ subalfred convert bytes2hex "[115, 99, 47, 107, 116, 115, 116, 107, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]"
0x73632f6b7473746b000000000000000000000000
```

File: pallet/staking/src/lib.rs [hunk 9]
```diff
@@ -1208,3 +1235,136 @@ where
...
+				internal_type: None,
+			}],
+			constant: None,
+			state_mutability: StateMutability::Payable,
+		};
+
+		let notify_transaction = LegacyTransaction {
+			nonce: U256::zero(),     // Will be reset in the message transact call
+			gas_price: U256::zero(), // Will be reset in the message transact call
+			gas_limit: U256::from(10_000_000), /* It should be big enough for the evm
```
[the remark this answers is not in the document]
Comment (hujw77) [reply pallet/staking/src/lib.rs:1335]: The impact is minimal, the actual weight used will not be much, but I believe 100W gas should be more than sufficient.

File: pallet/staking/src/lib.rs [hunk 10]
```diff
@@ -1307,21 +1307,21 @@ where
...
 				Param {
-					name: "reward".to_string(),
+					name: "reward".into(),
 					kind: ParamType::Uint(256),
 					internal_type: None,
 				},
 			],
 			outputs: vec![Param {
-				name: "success or not".to_string(),
+				name: "success or not".into(),
```
Comment (hackfisher) [inline pallet/staking/src/lib.rs:1324]: @boundless-forest Please review these change it is causing the CI fail.

**Review by hackfisher — CHANGES_REQUESTED** [review]

File: precompile/staking/src/lib.rs (hunk 6 above)
Comment (JayJay1024) [reply precompile/staking/src/lib.rs:100]: > @JayJay1024 Did we support restake on UI? Need to adapt if the answer is yes.

The current UI does not support restaked. Do we need to support restake on KTON Staking App?

File: pallet/staking/src/lib.rs (hunk 7 above)
Comment (AurevoirXavier) [reply pallet/staking/src/lib.rs:927]: Did you mean revert to `x/100` from `x/100+x`?

File: pallet/staking/src/lib.rs [hunk 11]
```diff
@@ -1208,3 +1235,132 @@ where
...
+
+		Perquintill::one() - Perquintill::from_rational(x, month_in_blocks)
+	}
+}
+/// The KTON reward curve during migrating to staking v2.
+pub fn migration_curve_kton_reward(x: Perquintill) -> Perquintill {
+	let x = x.deconstruct();
+
+	Perquintill::from_rational(x, x + Perquintill::one().deconstruct())
+}
```
Comment (hackfisher) [inline pallet/staking/src/lib.rs:1276]: Remove this.

File: pallet/staking/src/lib.rs [hunk 12]
```diff
@@ -883,42 +909,47 @@ pub mod pallet {
...
 		}
 
 		/// Distribute the session reward to staking pot and update the stakers' reward record.
-		pub fn distribute_session_reward(amount: Balance) -> Balance {
+		pub fn distribute_session_reward(amount: Balance) {
+			let (reward_to_v1, reward_to_v2) = {
+				let reward_to_ring = amount / 2;
+				let reward_to_kton = amount - reward_to_ring;
+				#[cfg(not(any(test, feature = "runtime-benchmarks")))]
+				let ratio = migration_curve_kton_reward(T::MigrationCurve::get());
```
Comment (hackfisher) [inline pallet/staking/src/lib.rs:924]:
```suggestion
				let ratio = T::MigrationCurve::get();
```

File: pallet/staking/src/lib.rs (hunk 7 above)
Comment (hackfisher) [reply pallet/staking/src/lib.rs:927]: Yes, using x directly, refer my suggestion above

File: precompile/staking/src/lib.rs (hunk 6 above)
Comment (hackfisher) [reply precompile/staking/src/lib.rs:100]: No need.

File: pallet/staking/src/lib.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (boundless-forest) [reply pallet/staking/src/lib.rs:1325]: @AurevoirXavier Please help adjust this to 100w gas~

**Review by hackfisher — CHANGES_REQUESTED** [review]

File: precompile/staking/src/lib.rs (hunk 6 above)
Comment (JayJay1024) [reply precompile/staking/src/lib.rs:100]: The current UI uses restake. I will adapt it.

File: pallet/staking/src/lib.rs [hunk 13]
```diff
@@ -883,42 +909,47 @@ pub mod pallet {
...
-				staker: staking_pot.clone(),
-				amount: actual_reward,
-			});
+			};
 
-			if unpaid != 0 {
-				Self::deposit_event(Event::Unpaid { staker: staking_pot, amount: unpaid });
-			}
+			reward(staking_pot, actual_reward_v1);
+			reward(T::KtonRewardDistributionOwner::get(), reward_to_v2);
```
[the remark this answers is not in the document]
Comment (hujw77) [reply pallet/staking/src/lib.rs:950]: Yes.

File: pallet/staking/src/lib.rs [hunk 14]
```diff
@@ -1208,3 +1235,126 @@ where
...
+			state_mutability: StateMutability::Payable,
+		};
+
+		let notify_transaction = LegacyTransaction {
+			nonce: U256::zero(),     // Will be reset in the message transact call
+			gas_price: U256::zero(), // Will be reset in the message transact call
+			gas_limit: U256::from(1_000_000), /* It should be big enough for the evm
+			                          * transaction, otherwise it will out of gas. */
+			action: TransactionAction::Call(reward_distr),
+			value: U256::zero(),
```
Comment (hackfisher) [inline pallet/staking/src/lib.rs:1328]: Since we pass zero value here, the value should be sent directly to reward_distr when reward.

**Review by boundless-forest — APPROVED** [review]
LGTM. Congrats!

File: precompile/metadata/abi/staking.json [hunk 15]
```diff
@@ -300,7 +294,7 @@
         "collect(uint32)": "10a66536",
         "nominate(address)": "b332180b",
         "payout(address)": "0b7e9c44",
-        "restake(uint256,uint256,uint8[])": "17092fcb",
+        "restake(uint256,uint8[])": "6dbcd550",
         "stake(uint256,uint256,uint8[])": "757f9b3b",
```
[the remark this answers is not in the document]
Comment (hackfisher) [reply precompile/metadata/abi/staking.json:298]: You mean close it on UI, Yes, I think we should. Just simply disable stake/restake KTON.

And provide a one click button to unstake all the KTON and guiding them to the new KTON Staking after runtime upgrade.

**Review by hujw77 — APPROVED** [review]
LGTM

**Pull request merged**